quadriga/src/main/java/edu/asu/diging/quadriga/web/service/impl/GraphCreationServiceImpl.java

```diff
@@ -0,0 +1,247 @@
+import java.util.Map;
+
+import org.bson.types.ObjectId;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import edu.asu.diging.quadriga.core.conceptpower.model.CachedConcept;
+import edu.asu.diging.quadriga.core.conceptpower.service.ConceptPowerService;
+import edu.asu.diging.quadriga.core.model.DefaultMapping;
+import edu.asu.diging.quadriga.core.model.EventGraph;
+import edu.asu.diging.quadriga.core.model.TripleElement;
+import edu.asu.diging.quadriga.core.model.elements.Relation;
+import edu.asu.diging.quadriga.core.model.events.AppellationEvent;
+import edu.asu.diging.quadriga.core.model.events.RelationEvent;
+import edu.asu.diging.quadriga.web.service.model.GraphData;
+import edu.asu.diging.quadriga.web.service.model.GraphEdgeData;
+import edu.asu.diging.quadriga.web.service.model.GraphElement;
+import edu.asu.diging.quadriga.web.service.model.GraphElements;
+import edu.asu.diging.quadriga.web.service.model.GraphNodeData;
+import edu.asu.diging.quadriga.web.service.model.GraphNodeType;
+import edu.asu.diging.quadriga.web.service.GraphCreationService;
+
+@Service
+public class GraphCreationServiceImpl implements GraphCreationService {
+
+    private Logger logger = LoggerFactory.getLogger(getClass());
+    
+    private final String source = "www.digitalhps.org";
+
+    @Autowired
+    private ConceptPowerService conceptPowerService;
+
+    @Override
+    public GraphElements createGraph(List<EventGraph> eventGraphs) {
+
+        List<GraphData> graphEdges = new ArrayList<>();
+        List<GraphData> graphNodes = new ArrayList<>();
+        Map<String, GraphNodeData> uniqueNodes = new HashMap<String, GraphNodeData>();
+
+        eventGraphs.stream()
+            .filter(eventGraph -> eventGraph.getRootEvent() instanceof RelationEvent)
+            .forEach(validEventGraph -> createNodesAndEdges((RelationEvent) (validEventGraph.getRootEvent()),
+                        graphNodes, graphEdges, uniqueNodes, validEventGraph.getId().toString()));
+
+        GraphElements graphElements = new GraphElements();
+        graphElements.setNodes(wrapInGraphElements(graphNodes));
+        graphElements.setEdges(wrapInGraphElements(graphEdges));
+
+        return graphElements;
+    }
+
+    @Override
+    public String createNodesAndEdges(RelationEvent event, List<GraphData> graphNodes, List<GraphData> graphEdges,
+            Map<String, GraphNodeData> uniqueNodes, String eventGraphId) {
+        Relation relation = event.getRelation();
+        String predicateNodeId = null;
+        String subjectNodeId = null;
+        String objectNodeId = null;
+
+        if (relation.getPredicate() != null) {
+            predicateNodeId = createPredicateNode(graphNodes, relation.getPredicate(), eventGraphId);
+        } else {
+            logger.error("A predicate is missing in one of the relations for EventGraph: " + eventGraphId);
```

Comment:
should the method return at this point?

Reply:
It can return as if there is no predicate an edge cant be formed.

Reply:
I don't understand what this means. Why create subject or object if there is no predicate?

Reply:
This method is being called by edu.asu.diging.quadriga.web.service.impl.GraphCreationServiceImpl i.e. for creating graphs. So, i see no point in just creating subject or object nodes without connecting edges. So should i return this function here?

Reply:
You are the one writing the code, but probably yes.

Reply:
this is not returning, but shouldn't it?